src/elm/external.ts

```diff
@@ -22,21 +25,49 @@ export class Retrieve extends Expression {
   }
 
   exec(ctx: Context) {
-    let records = ctx.findRecords(this.templateId != null ? this.templateId : this.datatype, this);
-    let codes = this.codes;
-    if (this.codes && typeof this.codes.exec === 'function') {
-      codes = this.codes.execute(ctx);
-      if (codes == null) {
+    // Object with retrieve information to pass back to patient source
+    const retrieveDetails: RetrieveDetails = {
+      datatype: this.datatype
+    };
+
+    let resolvedCodes: Code[] | ValueSet | undefined;
+    if (this.codes) {
+      resolvedCodes = this.codes.execute(ctx);
+
+      if (resolvedCodes == null) {
         return [];
       }
-    }
-    if (codes) {
-      records = records.filter((r: any) => this.recordMatchesCodesOrVS(r, codes));
+
+      retrieveDetails.codes = resolvedCodes;
     }
     // TODO: Added @dateProperty check due to previous fix in cql4browsers in cql_qdm_patient_api hash: ddbc57
+    let range: Interval | undefined;
     if (this.dateRange && this.dateProperty) {
-      const range = this.dateRange.execute(ctx);
-      records = records.filter((r: any) => range.includes(r.getDateOrInterval(this.dateProperty)));
+      range = this.dateRange.execute(ctx);
+
+      retrieveDetails.dateRange = range;
+      retrieveDetails.dateProperty = this.dateProperty;
+    }
+
+    if (this.templateId) {
+      retrieveDetails.templateId = this.templateId;
+    }
+
+    if (resolvedCodes && this.codeProperty) {
```

Comment:
And to continue on the theme, I don't think we need to guard this assignment with the check on `resolvedCodes`.  In fact, we might not need to guard it at all, because if `this.codeProperty` is not defined, there's no harm is assigning it to `retrieveDetails.codeProperty` anyway, right?

Reply:
I guess the only open question with this (and dateProperty), is if they're undefined, do we want them to be omitted entirely? Or show up on the object with a value of `undefined`?

Reply:
From a practical point of view, it probably doesn't matter.  But from my (sometimes perfectionist) point of view, I prefer they be omitted entirely.

But... do I prefer it strong enough to require an `if` block (3 lines of code) vs a single line of code?  Hmmm...  Tough question.

I will be happy either way.  Follow your heart, @mgramigna.